pkg/engine/events.go

```diff
@@ -112,7 +113,10 @@ type StepEventMetadata struct {
 	Provider string                  // the provider that performed this step.
 }
 
+// StepEventStateMetadata contains detailed metadata about a resource's state pertaining to a given step.
 type StepEventStateMetadata struct {
+	// State contains the raw, complete state, for this resource.
+	State *resource.State
```

Comment:
`StepEventStateMetadata` exists because we didn't want to pass the entire resource state out of the engine, but instead do some filtering beforehand (see https://github.com/pulumi/pulumi/blob/4b33a455615c8929ce79ee359fd18de83ebecd5d/pkg/engine/events.go#L203-L221). and the large commit above the `Inputs` field of this thing.

I see we are adding because we want to call SeralizeProperties to create a `ResourceV3`, but I'm wondering if we really want to commit to exposing all of this stuff across the channel.

How awful would it be to construct the new and old states inside the JSON render using just the existing data in the StepEventStateMetadata?

/cc @CyrusNajmabadi

Reply:
I personally believe it's better to leverage the existing data structures we plan to document, rather than forking and creating a shadow set that is kinda sorta the same, but not really.

Reply:
In that case, can we just remove all the other members of StepEventStateMetadata and move the filtering into the display layer for the other display sinks (or file an issue so we can make the cleanup later)?

Reply:
This also has some implications in how events are sent/persisted on the Pulumi Service. (Although adding a new field here won't cause problems, because we intentionally insulate the API type surface area from engine internals [here](https://github.com/pulumi/pulumi/blob/master/pkg/apitype/events.go#L94).)

But as we further integrate secrets and other things that require a more strict boundary between the "backend" processing the update, and the engine's internals, we need to be careful about things like this.

Reply:
Per my previous comment, these are the same structures we give to people in the checkpoint file, which they may already wish to tool. I don't see a reason to complicate the change. @ellismg I'm not entirely following your suggestion, so please feel free to file an issue.